// @transmute/rfc9162/dist/CoMETRE/RFC9162_SHA256.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RFC9162_SHA256 = void 0;
const getRootFromLeaves_1 = require("./getRootFromLeaves");
const getLeafFromEntry_1 = require("./getLeafFromEntry");
const getInclusionProofForLeaf_1 = require("./getInclusionProofForLeaf");
const getRootFromInclusionProof_1 = require("./getRootFromInclusionProof");
const getConsistencyProofFromLeaves_1 = require("./getConsistencyProofFromLeaves");
const verifyConsistencyProof_1 = require("./verifyConsistencyProof");
const leaves = (entries) => {
    return entries.map(getLeafFromEntry_1.getLeafFromEntry);
};
const root = (leaves) => {
    return (0, getRootFromLeaves_1.getRootFromLeaves)(leaves);
};
const iproof = (leaf_index, leaves) => {
    return (0, getInclusionProofForLeaf_1.getInclusionProofForLeaf)(leaf_index, leaves);
};
const viproof = (leaf, proof) => {
    return (0, getRootFromInclusionProof_1.getRootFromInclusionProof)(leaf, proof);
};
const cproof = (proof, leaves) => {
    return (0, getConsistencyProofFromLeaves_1.getConsistencyProofFromLeaves)(proof, leaves);
};
const vcproof = (first_root, second_root, proof) => {
    return (0, verifyConsistencyProof_1.verifyConsistencyProof)(first_root, proof.tree_size_1, second_root, proof.tree_size_2, proof);
};
const tree_alg = 'RFC9162_SHA256';
exports.RFC9162_SHA256 = {
    tree_alg,
    root,
    leaf: getLeafFromEntry_1.getLeafFromEntry,
    inclusion_proof: iproof,
    verify_inclusion_proof: viproof,
    consistency_proof: cproof,
    verify_consistency_proof: vcproof,
};
exports.default = exports.RFC9162_SHA256;
