// @transmute/verifiable-credentials/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.base64url = exports.validator = exports.verifier = exports.holder = exports.issuer = exports.text = exports.status = exports.claimset = exports.key = exports.version = void 0;
const jose_1 = require("jose");
Object.defineProperty(exports, "base64url", { enumerable: true, get: function () { return jose_1.base64url; } });
exports.version = 'https://www.w3.org/TR/2024/CR-vc-data-model-2.0-20240201/';
__exportStar(require("./types"), exports);
const key = __importStar(require("./key"));
exports.key = key;
const claimset = __importStar(require("./claimset"));
exports.claimset = claimset;
const status = __importStar(require("./status-list"));
exports.status = status;
const text = __importStar(require("./text"));
exports.text = text;
const credential_1 = require("./credential");
Object.defineProperty(exports, "issuer", { enumerable: true, get: function () { return credential_1.issuer; } });
const presentation_1 = require("./presentation");
Object.defineProperty(exports, "holder", { enumerable: true, get: function () { return presentation_1.holder; } });
const verifier_1 = require("./verifier");
Object.defineProperty(exports, "verifier", { enumerable: true, get: function () { return verifier_1.verifier; } });
const validator_1 = require("./validator");
Object.defineProperty(exports, "validator", { enumerable: true, get: function () { return validator_1.validator; } });
