// @transmute/cose/dist/cose/receipt/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifier = exports.remove = exports.get = exports.add = exports.consistency = exports.inclusion = exports.leaf = void 0;
const inclusion = __importStar(require("./inclusion"));
exports.inclusion = inclusion;
const consistency = __importStar(require("./consistency"));
exports.consistency = consistency;
const leaf_1 = require("./leaf");
Object.defineProperty(exports, "leaf", { enumerable: true, get: function () { return leaf_1.leaf; } });
const add_1 = require("./add");
Object.defineProperty(exports, "add", { enumerable: true, get: function () { return add_1.add; } });
const get_1 = require("./get");
Object.defineProperty(exports, "get", { enumerable: true, get: function () { return get_1.get; } });
const remove_1 = require("./remove");
Object.defineProperty(exports, "remove", { enumerable: true, get: function () { return remove_1.remove; } });
const verifier_1 = require("./verifier");
Object.defineProperty(exports, "verifier", { enumerable: true, get: function () { return verifier_1.verifier; } });
