// @transmute/cose/dist/cbor/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tagged = exports.diagnose = exports.decodeFirstSync = exports.decodeFirst = exports.encodeAsync = exports.decode = exports.encode = exports.encodeCanonical = exports.toArrayBuffer = exports.EMPTY_BUFFER = exports.Sign1Tag = void 0;
const cbor_web_1 = require("cbor-web");
Object.defineProperty(exports, "encodeCanonical", { enumerable: true, get: function () { return cbor_web_1.encodeCanonical; } });
Object.defineProperty(exports, "encode", { enumerable: true, get: function () { return cbor_web_1.encode; } });
Object.defineProperty(exports, "decode", { enumerable: true, get: function () { return cbor_web_1.decode; } });
Object.defineProperty(exports, "encodeAsync", { enumerable: true, get: function () { return cbor_web_1.encodeAsync; } });
Object.defineProperty(exports, "decodeFirst", { enumerable: true, get: function () { return cbor_web_1.decodeFirst; } });
Object.defineProperty(exports, "decodeFirstSync", { enumerable: true, get: function () { return cbor_web_1.decodeFirstSync; } });
Object.defineProperty(exports, "diagnose", { enumerable: true, get: function () { return cbor_web_1.diagnose; } });
Object.defineProperty(exports, "Tagged", { enumerable: true, get: function () { return cbor_web_1.Tagged; } });
const toArrayBuffer_1 = require("./toArrayBuffer");
Object.defineProperty(exports, "toArrayBuffer", { enumerable: true, get: function () { return toArrayBuffer_1.toArrayBuffer; } });
exports.Sign1Tag = 18;
exports.EMPTY_BUFFER = (0, toArrayBuffer_1.toArrayBuffer)(new Uint8Array());
